master/lustregittagpoller.py

```diff
@@ -0,0 +1,122 @@
+    to existing tags. Once a new or changed tag is found, a change is submitted to
+    the build master.""" 
+
+    def __init__(self, **kwargs):
+        if kwargs.get('workdir') is None:
+            kwargs['workdir'] = 'lustretagpoller-work'
+
+        GitPoller.__init__(self, **kwargs)
+
+    def describe(self):
+        str = ('LustreTagPoller: watching the remote git repository ' +
+               self.repourl + ' for new tags')
+
+        if not self.master:
+            str += " [STOPPED - check log]"
+
+        return str
+
+    def _getRefs(self):
+        """Collect and parse all remote references"""
+
+        d = self._dovccmd('ls-remote', [self.repourl])
+
+        @d.addCallback
+        def parseRefs(rows):
+            refs = {} 
+            for row in rows.splitlines():
+                if '\t' not in row:
+                    continue
+
+                sha, ref = row.split("\t")
+
+                if not self._filter_ref(ref):
+                    refs[ref] = sha
+
+            return refs
+
+        return d
+
+    def _filter_ref(self, ref):
+        """Filter refs that are not tags"""
+        return not ref.startswith("refs/tags/")
+
+    @defer.inlineCallbacks
+    def _process_change(self, ref, rev):
+        """Process each tag """
+        # initial run, don't parse all history
+        if not self.lastRev:
+            return
+
+        # if a tag has been previously seen, and it's revision hasn't changes, there's nothing to process
+        if ref in self.lastRev and self.lastRev[ref] == rev:
+            return
+
+        log.msg('LustreTagPoller: processing %s(%s) from "%s"'% (ref, rev, self.repourl))
+
+        dl = defer.DeferredList([
+            self._get_commit_timestamp(rev),
+            self._get_commit_author(rev),
+            self._get_commit_files(rev),
+            self._get_commit_comments(rev),
+        ], consumeErrors=True)
+
+        results = yield dl
+
+        # check for failures
+        failures = [r[1] for r in results if not r[0]]
+        if failures:
+            # just fail on the first error; they're probably all related!
+            raise failures[0]
+
+        timestamp, author, files, comments = [r[1] for r in results]
+        yield self.master.addChange(
+            author=author,
+            revision=rev,
+            files=files,
+            comments=comments,
+            when_timestamp=epoch2datetime(timestamp),
+            branch=ref,
+            category=self.category,
+            project=self.project,
+            repository=self.repourl,
+            src='git')
+
+    @defer.inlineCallbacks
+    def poll(self):
+        log.msg("LustreGitTagPoller: poll iteration started")
+        if not os.path.exists(self.workdir):
+            yield self._dovccmd('init', ['--bare', self.workdir])
+
+        log.msg("LustreGitTagPoller: collecting references")
+        newRefs = yield self._getRefs()
```

Comment:
ls-remote is an over-the-network operation.  After the very next step, the "fetch --tags", the information we need will be available locally, correct?  Can't we skip the ls-remote?